theano/gof/destroyhandler.py

```diff
@@ -394,21 +394,25 @@ def get_destroyers_of(r):
                 return []
         fgraph.destroyers = get_destroyers_of
 
-        def recursive_destroys_finder(clients_list):
-            for (app, idx) in clients_list:
+        def recursive_destroys_finder(clients_set):
+            for (app, idx) in clients_set:
                 if app == 'output':
                     continue
                 destroy_maps = getattr(app.op, 'destroy_map', {}).values()
                 if idx in [dmap for sublist in destroy_maps for dmap in sublist]:
                     return True
+                outputs_clients_set = set()
                 for var in getattr(app.op, 'view_map', {}).keys():
-                    if recursive_destroys_finder(app.outputs[var].clients):
-                        return True
+                    outputs_clients_set.update(app.outputs[var].clients)
+                if recursive_destroys_finder(outputs_clients_set):
+                    return True
             return False
 
         def has_destroyers(protected_list):
+            clients_set = set()
             for protected_var in protected_list:
-                if recursive_destroys_finder(protected_var.clients):
+                clients_set.update(protected_var.clients)
```

Comment:
So mostly, that commit don't help. There is some saving possible if one node reuse the same variable multiple time as input, but it happen so infrequently that I don't think the extra complexity is useful.